### PR TITLE
feat(rpc): add `eth_createAccessList` implementation

### PR DESCRIPTION
**Pull request by MegaRedHand**
Ref: https://github.com/paradigmxyz/reth/issues/1225

This PR adds an `eth_createAccessList` implementation.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 1]
```diff
@@ -222,11 +227,14 @@ where
         // transaction requires to succeed
         let gas_used = res.result.gas_used();
         // the lowest value is capped by the gas it takes for a transfer
-        let mut lowest_gas_limit = MIN_TRANSACTION_GAS;
+        let mut lowest_gas_limit =
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:230]: please keep the changes related to the PR,
it makes it harder to review if there are additional changes like this

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 2]
```diff
@@ -222,11 +227,14 @@ where
...
+            if (env.tx.transact_to.is_create()) { MIN_CREATE_GAS } else { MIN_TRANSACTION_GAS };
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:231]: style: redundant parentheses

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 3]
```diff
@@ -222,11 +227,14 @@ where
...
         let mut highest_gas_limit: u64 = highest_gas_limit.try_into().unwrap_or(u64::MAX);
         // pick a point that's close to the estimated gas
-        let mut mid_gas_limit =
-            std::cmp::min(gas_used * 3, (highest_gas_limit + lowest_gas_limit) / 2);
+        let mut mid_gas_limit = std::cmp::min(
+            gas_used * 3,
+            ((highest_gas_limit as u128 + lowest_gas_limit as u128) / 2) as u64,
+        );
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:237]: why is this better?

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 4]
```diff
@@ -269,25 +277,59 @@ where
...
+
+        // we want to disable this in eth_call, since this is common practice used by other node
+        // impls and providers <https://github.com/foundry-rs/foundry/issues/4388>
+        cfg.disable_block_gas_limit = true;
+
+        let mut env = build_call_evm_env(cfg, block, request)?;
+        let mut db = SubState::new(State::new(state));
+
+        let mut inspector = AccessListInspector::default();
+        let (_result, _env) = inspect(&mut db, env, &mut inspector)?;
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:302]: this should ensure the call succeeded?

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 5]
```diff
@@ -269,25 +277,59 @@ where
...
+        Ok(inspector.into_access_list())
+    }
 }
 
 /// Executes the [Env] against the given [Database] without committing state changes.
 pub(crate) fn transact<S>(db: S, env: Env) -> EthResult<(ResultAndState, Env)>
 where
     S: Database,
-    <S as Database>::Error: Into<EthApiError>,
+    <S as Database>::Error: Into<EthApiError> + std::fmt::Debug,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:312]: why is this required now?

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 6]
```diff
@@ -317,7 +359,7 @@ fn create_txn_env(block_env: &BlockEnv, request: CallRequest) -> EthResult<TxEnv
     let CallFees { max_priority_fee_per_gas, gas_price } =
         CallFees::ensure_fees(gas_price, max_fee_per_gas, max_priority_fee_per_gas)?;
 
-    let gas_limit = gas.unwrap_or(block_env.gas_limit);
+    let gas_limit = gas.unwrap_or(block_env.gas_limit.min(U256::from(u64::MAX)));
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:362]: why is this needed?
this seems redundant because the block env comes preconfigured?

File: crates/primitives/src/block.rs [hunk 7]
```diff
@@ -178,6 +178,12 @@ impl BlockId {
     }
 }
 
+impl Default for BlockId {
+    fn default() -> Self {
+        Self::Number(Default::default())
+    }
+}
```
Comment (mattsse) [inline crates/primitives/src/block.rs:185]: this will be a pitfall
please remove

File: crates/rpc/rpc/src/eth/api/server.rs [hunk 8]
```diff
@@ -190,10 +189,13 @@ where
...
-        _request: CallRequest,
-        _block_number: Option<BlockId>,
+        mut request: CallRequest,
+        block_number: Option<BlockId>,
     ) -> Result<AccessListWithGasUsed> {
-        Err(internal_rpc_err("unimplemented"))
+        let access_list = self.create_access_list_at(request.clone(), block_number).await?;
+        request.access_list = Some(access_list.clone());
+        let gas_used = self.estimate_gas_at(request, block_number.unwrap_or_default()).await?;
+        Ok(AccessListWithGasUsed { access_list, gas_used })
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/server.rs:198]: if block_number is none then this should be latest

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 9]
```diff
@@ -269,25 +277,59 @@ where
...
+        let state = self.state_at_block_id(at)?.ok_or_else(|| EthApiError::UnknownBlockNumber)?;
+
+        // we want to disable this in eth_call, since this is common practice used by other node
+        // impls and providers <https://github.com/foundry-rs/foundry/issues/4388>
+        cfg.disable_block_gas_limit = true;
+
+        let mut env = build_call_evm_env(cfg, block, request)?;
+        let mut db = SubState::new(State::new(state));
+
+        let mut inspector = AccessListInspector::default();
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:301]: we need to use the `new` function instead with the from and to address 

see

https://github.com/foundry-rs/foundry/blob/28b2ae6d5ad3e802b813b679e26bc15b9b8d230b/anvil/src/eth/backend/mem/mod.rs#L981-L994

**Review by mattsse — CHANGES_REQUESTED** [review]
ty for taking this on.

File: crates/primitives/src/block.rs (hunk 7 above)
Comment (MegaRedHand) [reply crates/primitives/src/block.rs:185]: Removed

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 3 above)
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/call.rs:237]: It accounts for possible overflow when `highest_gas_limit + lowest_gas_limit` is bigger than `u64::MAX`. This happened with the `CallRequest` used in `test_basic_eth_calls`. However, it could also be approximated by dividing both numbers before adding them (this would result in a maximum error of 2 units).

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 6 above)
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/call.rs:363]: https://github.com/paradigmxyz/reth/blob/a1d0b537a7549717956cb750d10045f353a9f8e6/crates/rpc/rpc/src/eth/cache.rs#LL225C49-L225C49

In this line, the cache initializes the `BlockEnv` to the default value, which uses `U256::MAX` for the gas limit. This results in an error when using the `NoopProvider` as `EvmEnvProvider`, as it doesn't modify this limit.

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 9 above)
Comment (onbjerg) [reply crates/rpc/rpc/src/eth/api/call.rs:302]: Why do we need an inspector to check this? The accounts read are accessible in the database (the first layer is just a cache DB, so it is populated every time something is read and starts out empty), and the written accounts are accessible in the changeset given by the executor after execution. I think it's more performant to just read these values versus using an inspector, since an inspector hooks into every opcode step, every call etc. (even if it does nothing there is still some overhead)

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 1 above)
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/call.rs:230]: Sorry, but these changes were needed for `test_basic_eth_calls` to pass. Should I move them to a separate PR?

This one in particular resulted in a _"intrinsic gas too high"_ error message because the gas limit passed to _revm_ was lower than the base cost of the transaction (which by default is a `TransactTo::Create`).

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 9 above)
Comment (mattsse) [reply crates/rpc/rpc/src/eth/api/call.rs:302]: ref https://github.com/foundry-rs/foundry/pull/2839

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 4 above)
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/call.rs]: Yes. Just added it.

File: crates/rpc/rpc/src/eth/api/call.rs (hunk 9 above)
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/call.rs]: Done

File: crates/rpc/rpc/src/eth/api/call.rs [hunk 10]
```diff
@@ -269,11 +279,77 @@ where
...
+                Err(InvalidTransactionError::Revert(RevertError::new(output)))
+            }
+            ExecutionResult::Success { .. } => Ok(()),
+        }?;
+        Ok(inspector.into_access_list())
+    }
+}
+
+/// Returns the addresses of the precompiles corresponding to the SpecId.
+fn get_precompiles(spec_id: &SpecId) -> Vec<reth_primitives::H160> {
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/call.rs:333]: this is fine for now, but we need this as part of the EvmProvider trait because precompiles are chain specific.

**Review by mattsse — APPROVED** [review]
tysm, I'll create a followup issue for the precompiles
